Double metric collection for tcp.connection.clientCount

`tcp.connection.clientCount` is collected and published twice as the assertion error reports.

```
AssertionError: metric '[unit=count,metric=tcp.connection.clientCount]' was rendered twice
    at com.hazelcast.internal.metrics.jmx.JmxPublisher.publishNumber(JmxPublisher.java:100)
    at com.hazelcast.internal.metrics.jmx.JmxPublisher.publishLong(JmxPublisher.java:86)
    at com.hazelcast.internal.metrics.impl.MetricsService$PublisherMetricsCollector.collectLong(MetricsService.java:250)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle.collectLong(MetricsCollectionCycle.java:165)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle.collect(MetricsCollectionCycle.java:144)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle.extractAndCollectDynamicMetrics(MetricsCollectionCycle.java:116)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle.access$300(MetricsCollectionCycle.java:47)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle$MetricsContext.collect(MetricsCollectionCycle.java:183)
    at com.hazelcast.internal.nio.tcp.TcpIpUnifiedEndpointManager.provideDynamicMetrics(TcpIpUnifiedEndpointManager.java:118)
    at com.hazelcast.internal.nio.tcp.TcpIpNetworkingService$MetricsProvider.provideDynamicMetrics(TcpIpNetworkingService.java:393)
    at com.hazelcast.internal.metrics.impl.MetricsCollectionCycle.collectDynamicMetrics(MetricsCollectionCycle.java:89)
    at com.hazelcast.internal.metrics.impl.MetricsRegistryImpl.collect(MetricsRegistryImpl.java:287)
    at com.hazelcast.internal.metrics.impl.MetricsService.collectMetrics(MetricsService.java:157)
    at com.hazelcast.internal.metrics.impl.MetricsService.collectMetrics(MetricsService.java:152)
    at com.hazelcast.spi.impl.executionservice.impl.DelegateAndSkipOnConcurrentExecutionDecorator$DelegateDecorator.run(DelegateAndSkipOnConcurrentExecutionDecorator.java:77)
    at com.hazelcast.internal.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:217)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
    at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/nio/tcp/TcpIpUnifiedEndpointManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/nio/tcp/TcpIpUnifiedEndpointManager.java
@@ -18,9 +18,6 @@ package com.hazelcast.internal.nio.tcp;
 
 import com.hazelcast.config.EndpointConfig;
 import com.hazelcast.instance.ProtocolType;
-import com.hazelcast.internal.metrics.MetricTagger;
-import com.hazelcast.internal.metrics.MetricTaggerSupplier;
-import com.hazelcast.internal.metrics.MetricsCollectionContext;
 import com.hazelcast.internal.metrics.Probe;
 import com.hazelcast.internal.networking.ChannelInitializerProvider;
 import com.hazelcast.internal.nio.IOService;
@@ -108,13 +105,5 @@ class TcpIpUnifiedEndpointManager
     @Probe(name = "textCount", level = MANDATORY)
     public int getCurrentTextConnections() {
         return getTextConnections().size();
-    }
-
-    @Override
-    public void provideDynamicMetrics(MetricTaggerSupplier taggerSupplier, MetricsCollectionContext context) {
-        super.provideDynamicMetrics(taggerSupplier, context);
-
-        MetricTagger tagger = taggerSupplier.getMetricTagger("tcp.connection");
-        context.collect(tagger, this);
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/nio/tcp/TcpIpEndpointManagerMetricsTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/nio/tcp/TcpIpEndpointManagerMetricsTest.java
@@ -1,0 +1,63 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.nio.tcp;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.metrics.collectors.MetricsCollector;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.OverridePropertyRule;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.test.OverridePropertyRule.set;
+import static com.hazelcast.test.TestEnvironment.HAZELCAST_TEST_USE_NETWORK;
+import static java.util.Collections.emptySet;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.matches;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class TcpIpEndpointManagerMetricsTest extends HazelcastTestSupport {
+
+    @Rule
+    public final OverridePropertyRule overridePropertyRule = set(HAZELCAST_TEST_USE_NETWORK, "true");
+
+    @Test
+    public void testUnifiedEndpointManagerMetricsCollectedOnce() {
+        MetricsCollector collectorMock = mock(MetricsCollector.class);
+        HazelcastInstance instance = createHazelcastInstance();
+
+        getNodeEngineImpl(instance).getMetricsRegistry().collect(collectorMock);
+
+        // defined by TcpIpEndpointManager
+        verifyCollectedOnce(collectorMock, "tcp.connection.count");
+        // defined by TcpIpUnifiedEndpointManager
+        verifyCollectedOnce(collectorMock, "tcp.connection.clientCount");
+        verifyCollectedOnce(collectorMock, "tcp.connection.textCount");
+    }
+
+    private void verifyCollectedOnce(MetricsCollector collectorMock, String metric) {
+        verify(collectorMock).collectLong(matches(".*" + metric + ".*"), anyLong(), eq(emptySet()));
+    }
+}

